### PR TITLE
several more ideas

### PR DESCRIPTION
**Pull request by reidpr**
Here are a bunch more ideas about how to change the spec. This is a translation of many of my critique items to specific change requests.

I'd like a feedback cycle or two before creating a bunch of Discussions.

### CODE REVIEW

File: _ideas/conformance_A3a.md [hunk 1]
```diff
@@ -0,0 +1,42 @@
+---
+title: Conformance tests
+discussion: FIXME
```
Comment (vsoch) [inline _ideas/conformance_A3a.md:3]: Don't worry about these - I'll add them all after it's merged.

File: _ideas/conformance_A3a.md [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+---
+
+## What are the challenges that we are trying to solve?
+
+No way to evaluate OCI conformance (aside from the [registry conformance test](https://github.com/opencontainers/distribution-spec/blob/main/conformance/README.md)).
+
+## What ideas do we have for how to address these challenges?
+
+Add conformance tests for image, runtime, and distribution client specs. These
```
Comment (vsoch) [inline _ideas/conformance_A3a.md:24]: This is a great idea! I think OCI would be very amenable to this.

File: _ideas/conformance_A3a.md [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+should report granular results rather than simply yes/no so that
+implementations can say what parts of OCI they do and don't support.
+
+## How do these ideas break down into...
+
+### Already existing OCI specs?
+
```
Comment (vsoch) [inline _ideas/conformance_A3a.md:31]:
```suggestion
This suggestion is relevant to the image, runtime, and a distribution client spec. Any news specs that arise should also be developed with conformance testing as a priority.
```

File: _ideas/content-length_D2.md [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+## What ideas do we have for how to address these challenges?
+
+Registries MUST return `Content-Length:` headers on all responses where it's
+meaningful, that match the manifest/config.
+
+## How do these ideas break down into...
+
+### Already existing OCI specs?
+
```
Comment (vsoch) [inline _ideas/content-length_D2.md:34]: This suggestion would be a change to the distribution spec.

File: _ideas/digest-input_B4.md [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+     purpose? And if they do have different purposes, the spec should explain
+     the reasoning.
+
+(The Charliecloud team had a lot of trouble debugging this because when a
+digest computation is wrong, you just get the wrong output with no indication
+of where the mistake is.)
+
+## What ideas do we have for how to address these challenges?
+
+Whenever a digest is mentioned, the spec should say right there there what the
```
Comment (vsoch) [inline _ideas/digest-input_B4.md:38]: What do you suggest for "right there" - another field? What would it look like?

File: _ideas/digest-input_B4.md [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+input is.
+
+## How do these ideas break down into...
+
+### Already existing OCI specs?
+
```
Comment (vsoch) [inline _ideas/digest-input_B4.md:44]:
```suggestion
This suggestion would be relevant to any interactions with [digests](https://github.com/opencontainers/go-digest).
```

File: _ideas/dockerfile_A6.md [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+It could be an opportunity to fix some of the inconsistencies, also; e.g.
+`ARG` and `ENV` do almost the same thing but have different syntax.
+
+## How do these ideas break down into...
+
+### Already existing OCI specs?
+
+
+### A new project (CNCF or OCI)
+
```
Comment (vsoch) [inline _ideas/dockerfile_A6.md:54]:
```suggestion
This idea suggests creating a new spec to describe the format of the Dockerfile. It should also come with conformance tests.
```

File: _ideas/entrypoint-cmd_B10.md [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+New instruction (`EXEC`?) that specifies both the command and default
+arguments.
+
+## How do these ideas break down into...
+
+### Already existing OCI specs?
+
+
+### A new project (CNCF or OCI)
+
```
Comment (vsoch) [inline _ideas/entrypoint-cmd_B10.md:38]:
```suggestion
If a Dockerfile spec could be created, this idea would supplement it.
```

File: _ideas/examples_A3b.md [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+Add edge case and complete examples to the specification text.
+
+There is some overlap here with "Conformance tests". The point here is
+slightly different, though: it's to illustrate for readers tricky parts of the
+spec.
+
+## How do these ideas break down into...
+
+### Already existing OCI specs?
+
```
Comment (vsoch) [inline _ideas/examples_A3b.md:38]:
```suggestion
This idea is relevant to all existing OCI specifications. Ideally there would be a standard means to define and display examples for the reader to quickly understand. If a documentation or similar working group is accepted this could be worked on.
```

File: _ideas/layer-diff_B2b.md [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+
+### A new project (CNCF or OCI)
+
+
+## Questions?
+
+
+## What pre-existing work or software can support these ideas?
+
+This scheme would work a lot like Git. Git does not store file metadata, but
```
Comment (vsoch) [inline _ideas/layer-diff_B2b.md:59]: That's such a cool idea. :)

File: _ideas/layer-diff_B2b.md [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+tarball layers in detail.
+
+## What ideas do we have for how to address these challenges?
+
+Store layers as binary diffs.
+
+## How do these ideas break down into...
+
+### Already existing OCI specs?
+
```
Comment (vsoch) [inline _ideas/layer-diff_B2b.md:49]:
```suggestion
This idea would be relevant for specs that are interested in interacting with an image manifest with layers.
```

File: _ideas/media-type_D1.md [hunk 12]
```diff
@@ -0,0 +1,44 @@
+---
+title: Honor media-type Accepts:
```
Comment (vsoch) [inline _ideas/media-type_D1.md:2]: If you want the `:` in the title we probably need quotes around it.
```suggestion
title: "Honor media-type Accepts:"
```

File: _ideas/media-type_D1.md [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+
+## What ideas do we have for how to address these challenges?
+
+Only return requested media types. If an implementation can accept multiple
+types, it can just say so with the `Accepts:` header.
+
+## How do these ideas break down into...
+
+### Already existing OCI specs?
+
```
Comment (vsoch) [inline _ideas/media-type_D1.md:33]:
```suggestion
This idea is relevant to the distribution spec.
```

File: _ideas/post-must_D4.md [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+
+## What ideas do we have for how to address these challenges?
+
+Require that registries support POST blob upload.
+
+
+## How do these ideas break down into...
+
+### Already existing OCI specs?
+
```
Comment (vsoch) [inline _ideas/post-must_D4.md:30]:
```suggestion
This idea is relevant to the distribution spec.
```

File: _ideas/runtime-unprivileged_C1.md [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+Split the configuration into required and optional components, and try to
+enforce that inclusion into required is only things that really are required.
+
+(To be clear, the latter will never fully be reliable, so Charliecloud would
+still allow overriding required things to work around config errors.)
+
+## How do these ideas break down into...
+
+### Already existing OCI specs?
+
```
Comment (vsoch) [inline _ideas/runtime-unprivileged_C1.md:43]:
```suggestion
This idea is relevant to the runtime spec.
```

File: _ideas/whiteouts-oob_B7.md [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+{: .no_toc .text-delta }
+
+1. TOC
+{:toc}
+
+---
+
+## What are the challenges that we are trying to solve?
+
+White-outs are in-band with special filenames, making it impossible to simply
```
Comment (vsoch) [inline _ideas/whiteouts-oob_B7.md:20]: Can you add a definition of what in/out of band means?

File: _ideas/conformance_A3a.md (hunk 1 above)
Comment (reidpr) [reply _ideas/conformance_A3a.md:3]: Great, thanks!

File: _ideas/digest-input_B4.md (hunk 5 above)
Comment (reidpr) [reply _ideas/digest-input_B4.md:38]: I tried again on this paragraph.

File: _ideas/whiteouts-oob_B7.md (hunk 16 above)
Comment (reidpr) [reply _ideas/whiteouts-oob_B7.md:20]: Yes, thanks.

**Comment by reidpr** [comment]
I think this is ready to merge if it looks OK to others.

**Comment by vsoch** [comment]
I think it looks good too - I'll merge and then update with the discussion URLs.
